src/main/java/org/gridsuite/modification/server/NetworkModificationService.java

```diff
@@ -8,75 +8,85 @@
 
 import com.powsybl.commons.PowsyblException;
 import com.powsybl.iidm.network.Network;
-import com.powsybl.iidm.network.Switch;
 import com.powsybl.network.store.client.NetworkStoreService;
 import groovy.lang.Binding;
 import groovy.lang.GroovyShell;
+import org.apache.commons.lang3.StringUtils;
 import org.codehaus.groovy.control.CompilerConfiguration;
-import org.gridsuite.modification.server.dto.GroovyScriptResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.annotation.ComponentScan;
-import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Service;
-import org.springframework.web.server.ResponseStatusException;
+import reactor.core.publisher.Mono;
 
 import java.util.Set;
 import java.util.UUID;
 
+import static org.gridsuite.modification.server.NetworkModificationException.Type.*;
+
 /**
  * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
  */
 @ComponentScan(basePackageClasses = {NetworkStoreService.class})
 @Service
 class NetworkModificationService {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(NetworkModificationService.class);
+
     @Autowired
     private NetworkStoreService networkStoreService;
 
-    private Network getNetwork(UUID networkUuid) {
-        try {
-            return networkStoreService.getNetwork(networkUuid);
-        } catch (PowsyblException e) {
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Network '" + networkUuid + "' not found");
-        }
+    private Set<String> doModification(Network network, Runnable modification) {
+        return doModification(network, modification, MODIFICATION_ERROR);
     }
 
-    public Set<String> changeSwitchState(UUID networkUuid, String switchId, boolean open) {
-        Network network = getNetwork(networkUuid);
-
-        Switch sw = network.getSwitch(switchId);
-        if (sw == null) {
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Switch " + switchId + " not found");
-        }
-
-        if (sw.isOpen() != open) {
-            sw.setOpen(open);
+    private Set<String> doModification(Network network, Runnable modification, NetworkModificationException.Type typeIfError) {
+        try {
+            DefaultNetworkStoreListener listener = new DefaultNetworkStoreListener();
+            network.addListener(listener);
+            modification.run();
 
             networkStoreService.flush(network);
 
-            return Set.of(sw.getVoltageLevel().getSubstation().getId());
-        } else {
-            return Set.of();
+            return listener.getModifications();
+        } catch (Exception e) {
+            NetworkModificationException exc = new NetworkModificationException(typeIfError, e.getMessage());
+            LOGGER.error(exc.getMessage());
```

Comment:
Here we lose the original stacktrace. Just add e as the cause of exc ?

Reply:
Done

Reply:
Done